File: ipc-codegen/echo_example/ts/golden_test.ts
/**
 * Golden file wire-format conformance test (TypeScript).
 *
 * For each golden file, asserts:
 *   1. We can decode the bytes into the expected typed value.
 *   2. Re-encoding the same value produces byte-identical output.
 * The combination pins down the wire format as a binding contract.
 *
 * Usage: npx tsx golden_test.ts
 */

import * as fs from "node:fs";
import * as path from "node:path";
import { Decoder, Encoder } from "msgpackr";

const decoder = new Decoder({ useRecords: false });
// `variableMapSize: true` makes msgpackr emit fixmap (1-byte header) for small
// maps instead of always reaching for map16. Without it the encoder produces
// a semantically-equivalent but byte-different encoding, so round-tripping
// the goldens would fail even though the wire is otherwise correct.
const encoder = new Encoder({ useRecords: false, variableMapSize: true });
const goldenDir = path.join(
  import.meta.dirname!,
  "../schema",
  "golden",
);

let pass = 0;
let fail = 0;

function bytesEqual(a: Uint8Array, b: Uint8Array): boolean {
  if (a.length !== b.length) return false;
  for (let i = 0; i < a.length; i++) {
    if (a[i] !== b[i]) return false;
  }
  return true;
}

function deepEqual(a: any, b: any): boolean {
  // For our test data: bigints, strings, numbers, plain arrays of u8 (which
  // msgpackr decodes as Uint8Array), and nested objects. The JSON-stringify
  // trick falls down on bigint and Uint8Array; do a structural walk.
  if (a === b) return true;
  if (typeof a === "bigint" || typeof b === "bigint") return a === b;
  if (a instanceof Uint8Array && b instanceof Uint8Array)
    return bytesEqual(a, b);
  if (Array.isArray(a) && Array.isArray(b)) {
    if (a.length !== b.length) return false;
    return a.every((x, i) => deepEqual(x, b[i]));
  }
  if (a && b && typeof a === "object" && typeof b === "object") {
    const ka = Object.keys(a).sort();
    const kb = Object.keys(b).sort();
    if (ka.length !== kb.length || !ka.every((k, i) => k === kb[i]))
      return false;
    return ka.every((k) => deepEqual(a[k], b[k]));
  }
  return false;
}

/** Read golden, decode, check expectation, and (optionally) verify re-encode
 *  byte-equals golden. Strict roundtrip is the binding wire-format check, but
 *  msgpackr has one known divergence from rmp-serde: positive bigints are
 *  encoded as int64 (`d3`) instead of uint64 (`cf`). Both encodings are
 *  accepted by every msgpack decoder we care about, so the wire is still
 *  interoperable — we just can't pin the bytes here. */
function check(
  file: string,
  expectedDecoded: any,
  opts: { strictRoundtrip?: boolean } = {},
) {
  const strictRoundtrip = opts.strictRoundtrip ?? true;
  try {
    const golden = fs.readFileSync(path.join(goldenDir, file));
    const decoded = decoder.unpack(golden);
    if (!deepEqual(decoded, expectedDecoded)) {
      throw new Error(
        `decoded mismatch:\n   got: ${stringify(decoded)}\n   exp: ${stringify(expectedDecoded)}`,
      );
    }
    if (strictRoundtrip) {
      const reencoded = encoder.encode(decoded);
      if (!bytesEqual(reencoded, golden)) {
        throw new Error(
          `roundtrip byte mismatch (decoded OK but re-encoded ${reencoded.length} bytes vs golden ${golden.length})`,
        );
      }
    }
    console.log(`  PASS: ${file}`);
    pass++;
  } catch (e: any) {
    console.log(`  FAIL: ${file}: ${e.message}`);
    fail++;
  }
}

function stringify(v: any): string {
  return JSON.stringify(v, (_k, x) => {
    if (typeof x === "bigint") return `${x}n`;
    if (x instanceof Uint8Array) return `[${Array.from(x).join(",")}]`;
    return x;
  });
}

console.log("Golden file wire-format conformance tests (TypeScript):\n");

// Request format: [[CommandName, {fields}]]
function req(cmdName: string, fields: any) {
  return [[cmdName, fields]];
}
// Response format: [ResponseName, {fields}]
function resp(respName: string, fields: any) {
  return [respName, fields];
}

// ============ Original happy-path cases ============
check(
  "echo_bytes_request.msgpack",
  req("EchoBytes", { data: new Uint8Array([0xde, 0xad, 0xbe, 0xef, 0x42]) }),
);
check(
  "echo_fields_request.msgpack",
  req("EchoFields", { a: 42, b: 999999, name: "hello wire compat" }),
);
check(
  "echo_nested_request.msgpack",
  req("EchoNested", {
    inner: {
      values: [new Uint8Array([1, 2, 3]), new Uint8Array([4, 5])],
      flag: true,
    },
  }),
);

check(
  "echo_bytes_response.msgpack",
  resp("EchoBytesResponse", {
    data: new Uint8Array([0xde, 0xad, 0xbe, 0xef, 0x42]),
  }),
);
check(
  "echo_fields_response.msgpack",
  resp("EchoFieldsResponse", { a: 42, b: 999999, name: "hello wire compat" }),
);
check(
  "echo_nested_response.msgpack",
  resp("EchoNestedResponse", {
    inner: {
      values: [new Uint8Array([1, 2, 3]), new Uint8Array([4, 5])],
      flag: true,
    },
  }),
);

// ============ Boundary cases ============

// bin8 empty + bin16 (256 bytes) — bin8/bin16 framing boundary.
check("echo_bytes_empty.msgpack", req("EchoBytes", { data: new Uint8Array() }));
check(
  "echo_bytes_bin16.msgpack",
  req("EchoBytes", { data: new Uint8Array(256).fill(0xaa) }),
);

// u32::MAX + u64::MAX + empty string. Largest uint encodings; fixstr-len-0.
// msgpackr decodes u64 fields as bigint when > Number.MAX_SAFE_INTEGER (2^53-1).
// Strict roundtrip is OK here because u64::MAX requires uint64 and msgpackr
// agrees with rmp-serde at the extreme.
check(
  "echo_fields_max.msgpack",
  req("EchoFields", { a: 4294967295, b: 18446744073709551615n, name: "" }),
);

// u32 = 128 (fixint → uint8 boundary), u64 above u32::MAX (forces uint64).
// strictRoundtrip: false — see check()'s comment about the bigint/uint64 quirk.
check(
  "echo_fields_uint_boundary.msgpack",
  req("EchoFields", { a: 128, b: 4294967296n, name: "x" }),
  { strictRoundtrip: false },
);

// Multi-byte UTF-8 in name.
check(
  "echo_fields_unicode.msgpack",
  req("EchoFields", { a: 0, b: 0, name: "héllo τέστ 🚀 mañana" }),
);

// 300-char ASCII string. Crosses fixstr (≤31) and str8 (≤255) into str16.
check(
  "echo_fields_str16.msgpack",
  req("EchoFields", { a: 0, b: 0, name: "a".repeat(300) }),
);

// Optional<bool> = absent (msgpackr decodes missing-with-nil to undefined or
// strips the key entirely depending on the encoder; rmp-serde emits a nil
// value for None inside a struct, so we expect flag: null here).
check(
  "echo_nested_flag_none.msgpack",
  req("EchoNested", { inner: { values: [], flag: null } }),
);

// Optional<bool> = Some(false) with values=[empty inner].
check(
  "echo_nested_flag_false.msgpack",
  req("EchoNested", { inner: { values: [new Uint8Array()], flag: false } }),
);

console.log(`\nResults: ${pass}/${pass + fail} passed, ${fail} failed`);
if (fail > 0) process.exit(1);

File: ipc-codegen/echo_example/cpp/src/golden_test.cpp
// Golden file wire-format conformance test (C++).
// For each golden file, asserts:
//   1. We can decode the bytes into the expected typed wire value.
//   2. Re-encoding the same value with the request/response framing produces
//      byte-identical output.
// The combination pins down the wire format as a binding contract.
//
// Usage: golden_test --golden-dir <path>

#include "generated/echo_dispatch.hpp"

#include <array>
#include <cstdint>
#include <cstring>
#include <fstream>
#include <iostream>
#include <iterator>
#include <optional>
#include <string>
#include <vector>

namespace {

int g_pass = 0;
int g_fail = 0;

std::vector<uint8_t> read_file(const std::string &path) {
  std::ifstream in(path, std::ios::binary);
  if (!in) {
    THROW std::runtime_error("cannot open " + path);
  }
  return std::vector<uint8_t>(std::istreambuf_iterator<char>(in),
                              std::istreambuf_iterator<char>());
}

void report(const std::string &file, const std::string &error) {
  if (error.empty()) {
    std::cerr << "  PASS: " << file << "\n";
    g_pass++;
  } else {
    std::cerr << "  FAIL: " << file << ": " << error << "\n";
    g_fail++;
  }
}

bool bytes_equal(const msgpack::sbuffer &buf,
                 const std::vector<uint8_t> &golden) {
  return buf.size() == golden.size() &&
         std::memcmp(buf.data(), golden.data(), golden.size()) == 0;
}

template <typename T> bool equals(const T &a, const T &b) { return a == b; }

bool equals(const echo::wire::EchoAliases &a,
            const echo::wire::EchoAliases &b) {
  return a == b;
}

bool equals(const echo::wire::EchoAliasesResponse &a,
            const echo::wire::EchoAliasesResponse &b) {
  return a == b;
}

// Requests are framed as [[name, payload-map]].
template <typename T>
void check_request(const std::string &dir, const std::string &file,
                   const std::string &name, const T &expected) {
  try {
    auto golden = read_file(dir + "/" + file);
    auto unpacked = msgpack::unpack(
        reinterpret_cast<const char *>(golden.data()), golden.size());
    auto obj = unpacked.get();
    if (obj.type != msgpack::type::ARRAY || obj.via.array.size != 1) {
      report(file, "expected outer array of size 1");
      return;
    }
    auto &inner = obj.via.array.ptr[0];
    if (inner.type != msgpack::type::ARRAY || inner.via.array.size != 2 ||
        inner.via.array.ptr[0].type != msgpack::type::STR) {
      report(file, "expected [CommandName, payload]");
      return;
    }
    std::string got_name(inner.via.array.ptr[0].via.str.ptr,
                         inner.via.array.ptr[0].via.str.size);
    if (got_name != name) {
      report(file, "wrong command name: " + got_name);
      return;
    }
    T value;
    inner.via.array.ptr[1].convert(value);
    if (!equals(value, expected)) {
      report(file, "decoded value mismatch");
      return;
    }
    msgpack::sbuffer buf;
    msgpack::packer<msgpack::sbuffer> pk(buf);
    pk.pack_array(1);
    pk.pack_array(2);
    pk.pack(name);
    pk.pack(value);
    if (!bytes_equal(buf, golden)) {
      report(file, "roundtrip byte mismatch (" + std::to_string(buf.size()) +
                       " vs " + std::to_string(golden.size()) + " bytes)");
      return;
    }
    report(file, "");
  } catch (const std::exception &e) {
    report(file, e.what());
  }
}

// Responses are framed as [name, payload-map].
template <typename T>
void check_response(const std::string &dir, const std::string &file,
                    const std::string &name, const T &expected) {
  try {
    auto golden = read_file(dir + "/" + file);
    auto unpacked = msgpack::unpack(
        reinterpret_cast<const char *>(golden.data()), golden.size());
    auto obj = unpacked.get();
    if (obj.type != msgpack::type::ARRAY || obj.via.array.size != 2 ||
        obj.via.array.ptr[0].type != msgpack::type::STR) {
      report(file, "expected [ResponseName, payload]");
      return;
    }
    std::string got_name(obj.via.array.ptr[0].via.str.ptr,
                         obj.via.array.ptr[0].via.str.size);
    if (got_name != name) {
      report(file, "wrong response name: " + got_name);
      return;
    }
    T value;
    obj.via.array.ptr[1].convert(value);
    if (!equals(value, expected)) {
      report(file, "decoded value mismatch");
      return;
    }
    msgpack::sbuffer buf;
    msgpack::packer<msgpack::sbuffer> pk(buf);
    pk.pack_array(2);
    pk.pack(name);
    pk.pack(value);
    if (!bytes_equal(buf, golden)) {
      report(file, "roundtrip byte mismatch (" + std::to_string(buf.size()) +
                       " vs " + std::to_string(golden.size()) + " bytes)");
      return;
    }
    report(file, "");
  } catch (const std::exception &e) {
    report(file, e.what());
  }
}

echo::wire::Fr test_hash(uint8_t base) {
  std::array<uint8_t, 32> bytes{};
  for (size_t i = 0; i < bytes.size(); i++) {
    bytes[i] = static_cast<uint8_t>(base + i);
  }
  return echo::wire::Fr(bytes);
}

} // namespace

int main(int argc, char **argv) {
  std::string dir;
  for (int i = 1; i + 1 < argc; i++) {
    if (std::string(argv[i]) == "--golden-dir") {
      dir = argv[i + 1];
    }
  }
  if (dir.empty()) {
    std::cerr << "Usage: golden_test --golden-dir <path>\n";
    return 1;
  }

  using namespace echo::wire;

  const std::vector<uint8_t> bytes_payload = {0xDE, 0xAD, 0xBE, 0xEF, 0x42};
  const EchoInner nested_inner{{{1, 2, 3}, {4, 5}}, true};
  const auto hash = test_hash(0x10);
  const auto second = test_hash(0x40);

  // ============ Original happy-path cases ============

  check_request(dir, "echo_bytes_request.msgpack", "EchoBytes",
                EchoBytes{bytes_payload});
  check_request(dir, "echo_fields_request.msgpack", "EchoFields",
                EchoFields{42, 999999, "hello wire compat"});
  check_request(dir, "echo_nested_request.msgpack", "EchoNested",
                EchoNested{nested_inner});
  check_request(dir, "echo_aliases_request.msgpack", "EchoAliases",
                EchoAliases{7, hash, second, {hash, second}});

  check_response(dir, "echo_bytes_response.msgpack", "EchoBytesResponse",
                 EchoBytesResponse{bytes_payload});
  check_response(dir, "echo_fields_response.msgpack", "EchoFieldsResponse",
                 EchoFieldsResponse{42, 999999, "hello wire compat"});
  check_response(dir, "echo_nested_response.msgpack", "EchoNestedResponse",
                 EchoNestedResponse{nested_inner});
  check_response(dir, "echo_aliases_response.msgpack", "EchoAliasesResponse",
                 EchoAliasesResponse{7, hash, second, {hash, second}});

  // ============ Boundary cases ============

  check_request(dir, "echo_bytes_empty.msgpack", "EchoBytes", EchoBytes{{}});
  check_request(dir, "echo_bytes_bin16.msgpack", "EchoBytes",
                EchoBytes{std::vector<uint8_t>(256, 0xAA)});
  check_request(dir, "echo_fields_max.msgpack", "EchoFields",
                EchoFields{UINT32_MAX, UINT64_MAX, ""});
  check_request(dir, "echo_fields_uint_boundary.msgpack", "EchoFields",
                EchoFields{128, uint64_t(UINT32_MAX) + 1, "x"});
  check_request(dir, "echo_fields_unicode.msgpack", "EchoFields",
                EchoFields{0, 0, "héllo τέστ 🚀 mañana"});
  check_request(dir, "echo_fields_str16.msgpack", "EchoFields",
                EchoFields{0, 0, std::string(300, 'a')});
  check_request(dir, "echo_nested_flag_none.msgpack", "EchoNested",
                EchoNested{EchoInner{{}, std::nullopt}});
  check_request(
      dir, "echo_nested_flag_false.msgpack", "EchoNested",
      EchoNested{EchoInner{std::vector<std::vector<uint8_t>>{{}}, false}});

  // ============ Blob / fail / error cases ============

  const std::array<std::vector<uint8_t>, 2> blob_parts = {
      std::vector<uint8_t>{1, 2, 3}, std::vector<uint8_t>{4}};
  const std::array<std::vector<uint8_t>, 2> blob_parts_none = {
      std::vector<uint8_t>{}, std::vector<uint8_t>{9}};

  check_request(dir, "echo_blobs_request.msgpack", "EchoBlobs",
                EchoBlobs{std::vector<uint8_t>{0xAA, 0xBB}, blob_parts});
  check_request(dir, "echo_blobs_none.msgpack", "EchoBlobs",
                EchoBlobs{std::nullopt, blob_parts_none});
  check_response(
      dir, "echo_blobs_response.msgpack", "EchoBlobsResponse",
      EchoBlobsResponse{std::vector<uint8_t>{0xAA, 0xBB}, blob_parts});
  check_request(dir, "echo_fail_request.msgpack", "EchoFail",
                EchoFail{"deliberate failure"});
  check_response(dir, "echo_fail_response.msgpack", "EchoFailResponse",
                 EchoFailResponse{});
  check_response(dir, "echo_error_response.msgpack", "EchoErrorResponse",
                 EchoErrorResponse{"deliberate failure"});

  std::cerr << "\nResults: " << g_pass << "/" << (g_pass + g_fail)
            << " passed, " << g_fail << " failed\n";
  return g_fail > 0 ? 1 : 0;
}
